Query evaluation: Add reciprocal rank

cbuescher opened the pull request:
This adds a second query evaluation metric alongside precision_at. Reciprocal Rank is defined as 1/rank, where rank is the position of the first relevant
document in the search result. The results are averaged across all queries across the sample of queries, according to
https://en.wikipedia.org/wiki/Mean_reciprocal_rank
This PR is against the WIP rank-eval feature branch and based on #19283.

Relates to #19195

cbuescher commented:
@MaineC opening this PR mostly so you can have a look, its WIP since it is based on #19283 which should be merged first. This only gives an impression about what other evalution metrics might look like.

MaineC commented on modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/ReciprocalRank.java:
```diff
@@ -0,0 +1,98 @@
+    public static final String NAME = "reciprocal_rank";
+    // the rank to use if the result list does not contain any relevant document
+    // TODO decide on better default or make configurable
+    private static final int RANK_IF_NOT_FOUND = Integer.MAX_VALUE;
```
I believe https://www.wikipendium.no/TDT4117_Information_retrieval has a better explanation of what to do if the document is not found at all.

The link above also states that the user should specify which is the largest acceptable rank. I think it makes sense to explicitly let users configure this: Depending on how many results are shown on the result page this number may vary. E.g. think of the difference of a mobile vs. a desktop result page: In the former case users are less likely to click results they have to scroll down to, while in the latter case those results might be readily visible without any user interaction.

cbuescher replied:
I will add a "largest acceptable rank" and use the description you linked to in the implementation for "not found".

MaineC commented:
Before merging I'd like to take a closer look at the original publication of the metric to make sure everything is fine. Other than that, looks good.

cbuescher commented:
@MaineC I added an optional parameter to the metric to specify the maximal allowed rank up to which the relevant first document is searched and added this to the unit and rest tests. Let me know what you think. btw. you mentioned an original publication, which would that be?

MaineC commented:
@cbuescher So, I found https://www.wikipendium.no/TDT4117_Information_retrieval#mean-reciprocal-rank-mrr which looks slightly different from what the wikipedia page. Now with this being just yet another wikipage I naively thought there must be an original paper somewhere - after digging for half an hour through references I don't find any. So scratch that comment of mine.

MaineC commented:
LGTM

cbuescher commented:
Jenkins, test this please.

cbuescher commented:
test this please
